### PR TITLE
[2.2.0] Rework on slippage and customisation

### PR DESCRIPTION
**Pull request by worlddlckgh**
- Reworked slippage 
  - support user's saved slippage in localstorage
  - fixed a bug where manual input, then clicking on preset, resulting in bad state
- Support useUserSlippage on init, integrator can choose to respect user's slippage or not
- Support initial slippage, allowing integrator to set an initial value, while respecting useUserSlippage or not

### CODE REVIEW

File: src/components/SwapSettingsModal/SwapSettingsModal.tsx [hunk 1]
```diff
@@ -95,7 +98,7 @@ const SetSlippage: React.FC<{ closeModal: () => void }> = ({ closeModal }) => {
   });
 
   /* SLIPPAGE */
-  const [inputFocused, setInputFocused] = useState(!slippageInitialPreset);
```
Comment (wkitsin) [inline src/components/SwapSettingsModal/SwapSettingsModal.tsx:98]: curious what was the bad state before

**Review by wkitsin — APPROVED** [review]

File: src/components/SwapSettingsModal/SwapSettingsModal.tsx [hunk 2]
```diff
@@ -193,13 +202,6 @@ const SetSlippage: React.FC<{ closeModal: () => void }> = ({ closeModal }) => {
       </div>
 
       <form
-        onSubmit={form.handleSubmit((value) => {
-          const slippage = Number(value.slippageInput ?? value.slippagePreset);
-          if (typeof slippage === 'number') {
-            setSlippage(slippage);
-            closeModal();
-          }
-        })}
```
Comment (ZhengYuTay) [inline src/components/SwapSettingsModal/SwapSettingsModal.tsx:202]: would prefer to onSubmit

File: src/contexts/SwapContext.tsx [hunk 3]
```diff
@@ -170,13 +182,29 @@ export const SwapContextProvider: FC<{
     () => ({ ...initialSwapContext.formProps, ...originalFormProps }),
     [originalFormProps],
   );
+  const [userSlippage, setUserSlippage] = useLocalStorage<number | undefined>('jupiter-terminal-slippage', undefined);
+  const [form, setForm] = useState<IForm>(
+    (() => {
+      const slippageBps = (() => {
+        if (props.useUserSlippage && typeof userSlippage !== 'undefined') {
```
Comment (ZhengYuTay) [inline src/contexts/SwapContext.tsx:189]: `typeof userSlippage !== 'undefined'` seems to be redundant

File: src/contexts/SwapContext.tsx [hunk 4]
```diff
@@ -126,6 +135,7 @@ export const initialSwapContext: ISwapContext = {
     priorityFeeInSOL: 0,
     setPriorityFeeInSOL() {},
   },
+  setUserSlippage: () => {},
```
Comment (ZhengYuTay) [inline src/contexts/SwapContext.tsx:138]: prefer to put this to `null` | `ISwapContext`

File: src/contexts/SwapContext.tsx [hunk 5]
```diff
@@ -220,19 +248,27 @@ export const SwapContextProvider: FC<{
...
-    } else {
-      if (!form.fromValue || !fromTokenInfo) return JSBI.BigInt(0);
-      return toLamports(Number(form.fromValue), Number(fromTokenInfo.decimals));
-    }
-  }, [form.fromValue, form.fromMint, fromTokenInfo, form.toValue, form.toMint, toTokenInfo, jupiterSwapMode]);
+  const jupiterParams: UseJupiterProps = useMemo(() => {
+    const amount = (() => {
+      if (jupiterSwapMode === SwapMode.ExactOut) {
+        if (!form.toValue || !toTokenInfo) return JSBI.BigInt(0);
+        return JSBI.BigInt(toLamports(Number(form.toValue), Number(toTokenInfo.decimals)));
```
Comment (ZhengYuTay) [inline src/contexts/SwapContext.tsx:255]: safe to do `Number` here?

File: src/index.html [hunk 6]
```diff
@@ -3,7 +3,7 @@
   <head>
     <meta charset="utf-8" />
     <title>Jupiter Terminal</title>
-    <script src="https://jupiter-terminal-git-v2-wowcats.vercel.app/main-v2.js" data-preload></script>
+    <script src="http://localhost:3000/main-v2.js" data-preload></script>
```
Comment (ZhengYuTay) [inline src/index.html:6]: rmb to update this

File: src/components/SwapSettingsModal/SwapSettingsModal.tsx (hunk 1 above)
Comment (worlddlckgh) [reply src/components/SwapSettingsModal/SwapSettingsModal.tsx:98]: Down to race condition between onchange, and react's batching of renders

File: src/contexts/SwapContext.tsx (hunk 3 above)
Comment (worlddlckgh) [reply src/contexts/SwapContext.tsx:189]: userSlippage can be undefined if the user have not changed their slippage before, so it's required

File: src/contexts/SwapContext.tsx (hunk 4 above)
Comment (worlddlckgh) [reply src/contexts/SwapContext.tsx:138]: can you clarify what this means?

File: src/contexts/SwapContext.tsx (hunk 5 above)
Comment (worlddlckgh) [reply src/contexts/SwapContext.tsx:255]: definitely not

File: src/index.html (hunk 6 above)
Comment (worlddlckgh) [reply src/index.html:6]: this is fine, for local testing usage

**Review by ZhengYuTay — APPROVED** [review]

**Pull request merged**